### PR TITLE
Update for 1.25

### PR DESCRIPTION
**Pull request by starlm**
Should work on 1.20 and newer but I haven't tested anything below 1.24.
I've been using it for a week now and encountered no problems.

Arcs & chains support is untested as of right now but it should work.

Edit: Arcs & chains support is implemented in a way to not bring the average down, right now it just sets it to 1.0 if it's not being rated by the game

### CODE REVIEW

**Comment by kinsi55** [comment]
I will try to look into this some time soon thank you

**Comment by starlm** [comment]
Just wanted to give a quick update on 1.20+ support. 
I recently finished an implementation for measuring overswing on arcs and chains reliably. 
Branch: [OrdinaryTable:arc-chain-support](https://github.com/OrdinaryTable/BeatSaber_OverswingCounter/tree/arc-chain-support)

In case you want them to be measured, I could merge it into [OrdinaryTable:1.25](https://github.com/OrdinaryTable/BeatSaber_OverswingCounter/tree/1.25).

I've been using it for some weeks now with no problems and no changes to the actual score, as that would of course be quite bad.

Note: Implementation could technically be simplified by a lot but I'm worried that it might end up changing the score in certain scenarios

**Comment by kinsi55** [comment]
Hey, thanks for the update and sorry that I didnt get around to looking into this yet. I think arcs and chains should just be ignored unless explicitly enabled (opt-in)

 I'll compile this PR and try it next time I play

**Comment by kinsi55** [comment]
Neat to see that it supposedly works in replays now? Did you validate that? Based of a quick look and knowing my left hand the values dont look completely off hence me asking :D

**Comment by starlm** [comment]
It does work in replays but only in certain scenarios, the play style (i.e. swing speed, etc.) heavily affects the accuracy of the overswing. When comparing it to live data it's slightly off.

**Comment by starlm** [comment]
5f52d36 fixed arcs & chains not being rated in a certain scenario.

I would also suggest completely disabling the counter when watching a ScoreSaber/BeatLeader replay, as SS replays result in inaccurate data and BL replays don't provide any ratings at all (i.e. 0% pre and 0% post)

**Comment by kinsi55** [comment]
Probably just display - for the values in those cases yeah

**Comment by kinsi55** [comment]
I was gonna go over all your transpilers but I'm starting to wonder if, instead of transpiling the game to hell and back, it wouldnt instead make sense to instead just seperately track the saber movements and calculate the angles manually on note cut events.

This way it would also properly work in replays, and be much less prone to breaking on updates

**Comment by starlm** [comment]
It might be a better solution but we would need to do all of the checks which the game does ourselves, i.e. when did the saber reach the center of the note, when does the swing actually end, etc.

I highly doubt that BeatGames will do any changes to how cuts are calculated any time soon, as the only thing which actually broke in 1.20, even tho it changed quite a lot, was the way it created a new CutInfo instance.

But at the end it's really a question of how much more complicated or prone to breaking it would be.

**Comment by kinsi55** [comment]
> i.e. when did the saber reach the center of the note, when does the swing actually end, etc.

This actually isnt a thing - Angle calculation is extremely primitive. Preswing is the angle up until the point when you touch the note hitbox, and postswing is everything after

**Comment by starlm** [comment]
I see. If we do it your way, it would probably be best to just completely rewrite it from scratch (or at least everything but the custom counter).

**Comment by kinsi55** [comment]
Yeah, that would make the most sense. It could reuse a lot of the basegame components that are used for angle calculation, just with adjusted logic / removed clamping (Methods can be generated at runtime with copied / adjusted IL)

I'll probably still test / release your current solution so that something does exist for now and leave this as something to worry about later :tm:

**Comment by starlm** [comment]
Sounds good. I'll get started on the rewrite in the mean time, if that's okay with you.

**Comment by kinsi55** [comment]
If you wanna look into that sure, be my guest

**Comment by starlm** [comment]
Just wanted to give a quick update on how the rewrite is going.
It's coming along very well and so far it has 0 transpilers and 1 prefix (I'm not expecting that to change).
By the way it works in ScoreSaber and BeatLeader replays, I have not tested the accuracy yet but it looks pretty good.

I'll give you another update once it's done (Presumingly sometime next week).

**Comment by kinsi55** [comment]
Thats about what I would have expected, sounds great!

**Comment by starlm** [comment]
I looked into replay support and I don't think that accurate data will be possible. Due to beat saber adding new data to the ``SaberMovementData`` instance each frame, it simply won't be possible unless the frame rate matches the replay's frame rate.

That being said, it certainly is far more unlikely to break with updates and more accurate now but that's about it.

**Comment by kinsi55** [comment]
Even if its not accurate, it should be a good enough estimate I'd guess - All of this really doesnt need to be extremely accurate as its really only meant to give you an idea how close you are to underswinging / how much you are overswinging / what your balance looks like

**Comment by starlm** [comment]
I should've worded it differently, sometimes it's a good 20-40° off. But just like before, it's heavily affected by the play style.
I'll try to make a comparison video later today to show how far off it is compared to live data.

**Comment by kinsi55** [comment]
Dont need to worry about that, I'll see for myself whenever I happen to test it 👍

**Comment by starlm** [comment]
Rewrite is done. Overswing appears to be pretty accurate. In terms of replays, ScoreSaber performed way better than expected. BeatLeader on the other hand, performed horrendously bad.

I made a little side-by-side comparison video: https://youtu.be/V0UNAkmzXPU

If you want to try it yourself, I could send you the source code over Discord or create a new branch, whatever you prefer.

**Comment by kinsi55** [comment]
You could just commit it into this PR's branch

I wonder why it is so bad on Beatleader - I have a feeling its down to pulling the values at the wrong frame seeing as consistently the upper value is bigger by about the amount missing at the bottom.

File: Models/DataProcessor.cs [hunk 1]
```diff
@@ -50,6 +50,8 @@ public class DataProcessor : ISaberMovementDataProcessor {
 				var beforeCutBottomPos = prevData.bottomPos;
 				var afterCutTopPos = newData.topPos;
 				var afterCutBottomPos = newData.bottomPos;
+
+				// Raycasts are somewhat expensive - Might make sense to reuse the SSRC's values with a transpiler
```
Comment (starlm) [inline Models/DataProcessor.cs:55]: I looked into it and it tended to be 3-5x slower. When reusing the values, it sometimes didn't get them for some reason but it reused 85-90% of SSRC's values.
It's mostly slower due to it making sure that the newest ``BladeMovementDataElement`` is the same to not mess up overswing.

Comment (kinsi55) [reply Models/DataProcessor.cs:55]: Weird - I guess your transpiler just stored the result in some static field that you then retrieved?

Comment (starlm) [reply Models/DataProcessor.cs:55]: It added the ray result into a dictionary (`Dictionary<SaberSwingRatingCounter, Dictionary<BladeMovementDataElement, Vector3>>`), `DataProcessor` then tries to get the result from it. It sounds very inefficient, I know, but not using the SSRC and `BladeMovementDataElement` to get the correct one, completely messed up the overswing.

Comment (kinsi55) [reply Models/DataProcessor.cs:55]: Shouldnt the singlethreaded nature mean that the basegame `ProcessNewData` is always executed before the counters one is, and thus you'd only need the "last" value?

Comment (starlm) [reply Models/DataProcessor.cs:55]: That's what I tried before, it tended to double the overswing.

Comment (kinsi55) [reply Models/DataProcessor.cs:55]: Hm, well, realisticly it shouldnt really matter, was just a possible optimization that came to mind

**Comment by kinsi55** [comment]
[Played with it yesterday](https://www.twitch.tv/videos/1639708051). I didnt pay all that much attention to it but generally the numbers did not seem to be "obviously wrong". I'll try give it a more technical test some time soon

**Comment by kinsi55** [comment]
Small update, really sorry I didnt merge / release this yet. I've had this plugin installed for all that time now its just that every time I play I forget to actually give it a deeper look :D I'll try to remember next time I play

**Comment by kinsi55** [comment]
Looked at it today briefly, seems correct at first glance. I'll be releasing this in a bit 👍

**Pull request merged**